server/routerlicious/packages/services-shared/src/redisSocketIoAdapter.ts

```diff
@@ -0,0 +1,506 @@
+     */
+    private async unsubscribeFromRooms(rooms: string[]) {
+        await RedisSocketIoAdapter.options.subConnection.unsubscribe(this.getChannelNames(rooms));
+
+        for (const room of rooms) {
+            this.clearRoomHealthCheckTimeout(room);
+        }
+    }
+
+    private getChannelNames(rooms: string[]) {
+        return rooms.map((room) => `${this.channel}${room}#`);
+    }
+
+    /**
+     * Queues a future health check
+     */
+    private queueRoomHealthCheck(room: string) {
+        this.clearRoomHealthCheckTimeout(room);
+
+        if (!RedisSocketIoAdapter.options.healthChecks ||
+            !RedisSocketIoAdapter.options.subConnection.isSubscribed(room)) {
+            return;
+        }
+
+        const timeoutId = setTimeout(
+            // eslint-disable-next-line @typescript-eslint/no-misused-promises
+            async () => {
+                return this.runRoomHealthCheck(room);
+            },
+            RedisSocketIoAdapter.options.healthChecks.interval);
+        this.roomHealthCheckTimeoutIds.set(room, timeoutId);
+    }
+
+    /**
+     * Runs a health check
+     * It will publish a message over the pub connection and wait until it receives it
+     */
+    private async runRoomHealthCheck(room: string) {
+        const healthCheckId = uuid.v4();
+
+        const startTime = Date.now();
+
+        const callerId = `${room},${healthCheckId}`;
+
+        try {
+            const msg = msgpack.encode([this.uid, healthCheckId]);
+
+            const healthCheckPromise = new Promise<void>((resolve) => {
+                this.pendingHealthChecks.set(healthCheckId, resolve);
+            });
+
+            // tslint:disable-next-line: no-floating-promises
+            await RedisSocketIoAdapter.options.pubConnection.publish(`${this.channel}${room}#`, msg);
+
+            await promiseTimeout(RedisSocketIoAdapter.options.healthChecks.timeout, healthCheckPromise);
+
+            if (RedisSocketIoAdapter.options.healthChecks.onHealthCheck) {
+                RedisSocketIoAdapter.options.healthChecks.onHealthCheck(callerId, startTime);
+            }
+        } catch (ex) {
+            if (RedisSocketIoAdapter.options.healthChecks.onHealthCheck) {
+                RedisSocketIoAdapter.options.healthChecks.onHealthCheck(callerId, startTime, ex);
+            }
+
+            if (RedisSocketIoAdapter.options.healthChecks.resubscribeOnFailure) {
+                // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                this.subscribeToRooms([room]);
+            }
+        } finally {
+            this.pendingHealthChecks.delete(healthCheckId);
+        }
+
+        // queue a health check even though we are not currrently subscribed
+        // the fact that this health check timer is still running means we still want to be subscribed to the room
+        // likely caused by a redis disconnection & a reconnection is in progress
+        this.queueRoomHealthCheck(room);
```

Comment:
It looks like queueRoomHealthCheck() calls runRoomHealthCheck(), which in turn calls queueRoomHealthCheck() again - it looks recursive. If that's the case, is there any chance the recursion never ends and we run out of resources or something similar? Apologies if I'm missing something obvious :)

Reply:
It is recursive but it will eventually end. When all clients leave a room, it triggers an unsubscribe call for that channel. And so the `queueRoomHealthCheck` method will stop because `RedisSocketIoAdapter.options.subConnection.isSubscribed(room)` will return `false`